introduction.Rmd

```diff
@@ -184,6 +184,309 @@ We call the combination of these two steps the __quote and unquote__ pattern. Th
+
+
+## Automating loops {#sec:automating-loops}
+
+Most programming problems involve __iteration__ because data transformations are typically achieved element by element, by applying the same recipe over and over again. There are two main ways of automating iteration in R, __vectorisation__ and __mapping__. Learning how to juggle with the different ways of expressing loops is not only an important step towards acquiring a good command of R and the tidyverse, it will also make you more proficient at solving programming problems.
+
+
+### Vectorisation in dplyr
+
+dplyr is designed to optimise iteration by taking advantage of the vectorisation of many R functions. Rowwise vectorisation is achieved through normal R rules, which dplyr augments with groupwise vectorisation.
+
+
+#### Rowwise vectorisation
+
+Rowwise vectorisation in dplyr is a consequence of normal R rules for vectorisation. A vectorised function is a function that works the same way with vectors of 1 element than with vectors of _n_ elements. The operation is applied elementwise (often at the machine code level, which makes them very efficient). We have already mentioned about the vectorisation of `toupper()`, and many other functions in R are vectorised. One important class of vectorised functions are the arithmetic operators:
+
+```{r}
+# Dividing 1 element
+1 / 10
+
+# Dividing 5 elements
+1:5 / 10
+```
+
+Technically, a function is vectorised when:
+
+* It returns a vector as long as the input.
+* Applying the function on a single element yields the same result than applying it on the whole vector and then subsetting the element.
+
+In other words, a vectorised function `fn` fulfills the following identity:
+
+```{r, eval = FALSE}
+fn(x[[i]]) == fn(x)[[i]]
+```
+
+When you mix vectorised and non-vectorised operations, the combined operation is itself vectorised when the last operation to run is vectorised. Here we'll combine the vectorised `/` function with the summary function `mean()`. The result of this operation is a vector that has the same length as the LHS of `/`:
+
+```{r}
+x <- 1:5
+x / mean(x)
+```
+
+Note that the other combination of operations is not vectorised because in that case the summary operation has the last word:
+
+```{r}
+mean(x / 10)
+```
+
+The dplyr verb `mutate()` expects vector semantics. The operations defining new columns typically return vectors as long as their inputs:
+
+```{r}
+data <- tibble(x = rnorm(5, sd = 10))
+
+data %>%
+  mutate(rescaled = x / sd(x))
+```
+
+In fact, `mutate()` enforces vectorisation. Returning smaller vectors is an error except if they have size 1. If the result of a mutate expression is a constant, it is automatically recycled to the tibble or group size. This ensures that all columns have the same length and fit within the tibble constraints of rectangular data:
```

Comment:
* swap in unless for "except"? 
* Is there a reason to prefer "size" to "length" here?

Reply:
"Size" is the term used in vctrs for rowwise length, for instance `vctrs::vec_size(mtcars)` is 32 while its `length()` is 11.

Reply:
Gotcha. I thought it might be vctrs related. Resolved!